design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+This approach has many drawbacks, here are a few examples:
+
+* We cannot discover database failures
+* We cannot discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address these problems, we need provide a parametrized Healthcheck endpoint in the Server and use it in our
+container image as well as in the Operator.
+
+## Parameterizing the Healthcheck URL
+
+The new Health Endpoint need to support query parameters in the endpoint in the following form:
+
+    http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google
+
+The following list of additional parameters will be supported:
+
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures.
+  * `idp-<identity-provider-alias>` - failing when an IdP is not accessible.
+  * `user-federation-<id>` - failing is LDAP connection is down.
+* (optional) `timeout` - the request gets terminated if not performed within a timeout.
+The implementation needs to take special care of releasing the thread pool (if necessary).
+Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly. If not specified,
+`10s` will be used as default (the same timeout as the one used in Kubernetes).
+
+## Implementation
+
+From the implementation perspective, there are two ways of implementing Healcheck calls within Keycloak:
+* Active - whenever we check if a database is down (this is just an example), we initiate a new connection and
+check if it succeeds.
+* Passive - we position the implementation based on Metrics components (see the Observability Design). We actively
+analyze database connection metrics reported by JCA subsystem
+(see [the currently available metrics in Keycloak](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit#gid=0))
+and if we discover  increasing failure count, we report an at Health Endpoint as well.
+
+The former approach seems more natural but has a lof of drawbacks, including:
+* Additional connections allocated in the thread pools.
+* Often requires integrating with additional subsystems (e.g. JCA for database checks).
+* Complicated to ensure that `timeout` parameter is properly respected.
+* Probing External IdPs might be complicated (e.g. what endpoint to call? what about authentication?)
```

Comment:
Actually, for the db check it should be active I'd say, but active part is done by the datasource, where it should keep at least one connection, and have a check on the connection periodically. This would mean that an active Keycloak would just rely on regular db requests, while an idle Keycloak would have a "ping" to the DB.

Reply:
I believe I answered this here: https://github.com/keycloak/keycloak-community/pull/55#discussion_r496539094

Reply:
From the design here you are stating that DB checks is passive, meaning it will require traffic to Keycloak to detect DB is down. However, in your answer you linked you are saying that it is basically active, since it is keeping one active connection in the pool. So that means the design doesn't reflect the actual plan.

Reply:
Yes, perhaps I should clarify this a bit more.

The Passive Mode in this context means that we don't actively force JCA subsystem to check the connections. We just observe the metrics it emits. But on the other hand, the JCA subsystem actively maintains the connection pool and validates connections inside of it.

I'll try to make it a bit more clear in the design document.

Reply:
Clarified in the new version. Resolving.